secio/src/codec/secure_stream.rs

```diff
@@ -297,100 +89,162 @@ where
+                    Ok(n)
+                } else {
+                    // fill internal recv buffer
+                    self.recv_buf = decoded;
+                    // drain for input buffer
+                    let copied = self.drain(buf);
+                    Ok(copied)
+                }
+            }
+            Some(Err(err)) => {
+                return Err(err.into())
+            }
+            None => {
+                debug!("connection shutting down");
+                return Err(io::ErrorKind::BrokenPipe.into())
+            }
         }
+    }
 
-        if !self.pending.is_empty() || !self.read_buf.is_empty() {
-            self.flush(cx)?;
+    #[inline]
+    fn encode_buffer(&mut self, buf: &[u8]) -> Bytes {
+        let mut out = self.encode_cipher.encrypt(buf).unwrap();
+        if let Some(ref mut hmac) = self.encode_hmac {
+            let signature = hmac.sign(&out[..]);
+            out.extend_from_slice(signature.as_ref());
         }
+        Bytes::from(out)
+    }
 
-        self.poll_complete(cx)?;
+    async fn write_socket(&mut self, buf: &[u8]) -> io::Result<usize> {
+        debug!("start sending plain data: {:?}", buf);
+        let l = buf.len();
 
-        match self.recv_frame(cx) {
-            Poll::Ready(Ok(None)) => {
-                self.close();
-                return Poll::Ready(None);
+        let frame = self.encode_buffer(buf);
+        trace!("start sending encrypted data size: {:?}", frame.len());
+        match self.socket.send(frame).await {
+            Ok(()) => {
+                Ok(l)
             }
-            Poll::Ready(Err(err)) => {
-                debug!("receive frame error: {:?}", err);
-                self.close();
-                return Poll::Ready(Some(Err(err.into())));
+            Err(err) => {
+                return Err(err.into())
             }
-            _ => (),
         }
+    }
+}
+
+impl<T> AsyncRead for SecureStream<T>
+    where
+        T: AsyncRead + AsyncWrite + Unpin,
+{
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        poll_future(cx, self.read_socket(buf))
+    }
+}
 
-        self.recv_event(cx);
+impl<T> AsyncWrite for SecureStream<T>
+    where
+        T: AsyncRead + AsyncWrite + Unpin,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        poll_future(cx, self.write_socket(buf))
+    }
 
-        // Double check stream state
-        if self.dead && self.nonce.is_empty() {
-            return Poll::Ready(None);
-        }
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        poll_future(cx, self.socket.flush())
+    }
 
-        Poll::Pending
+    fn poll_shutdown(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
```

Comment:
should poll shutdown inner socket

Reply:
changed to poll_future(cx, self.socket.close())